File: 329. Longest Increasing Path in a Matrix.py
class Solution(object):
    def longestIncreasingPath(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: int
        """
        from copy import deepcopy
        if not matrix:
            return 0
        m = len(matrix)
        n = len(matrix[0])
        inx = [[0] * n for i in range(m)]
        out = [[0] * n for i in range(m)]
        cnt = [[1] * n for i in range(m)]
        for i in range(m):
            for j in range(n):
                if i-1>=0 and matrix[i][j] < matrix[i-1][j]:
                    inx[i-1][j] += 1
                    out[i][j] += 1
                if i+1<m and matrix[i][j] < matrix[i+1][j]:
                    inx[i+1][j] += 1
                    out[i][j] += 1
                if j-1>=0 and matrix[i][j] < matrix[i][j-1]:
                    inx[i][j-1] += 1
                    out[i][j] += 1
                if j+1<n and matrix[i][j] < matrix[i][j+1]:
                    inx[i][j+1] += 1
                    out[i][j] += 1
        q1 = []
        for i in range(m):
            for j in range(n):
                if inx[i][j] == 0:
                    q1.append([i, j])
        
        ret = 0
        while len(q1) > 0:
            t = q1.pop()
            i, j = t
            ret = max(ret, cnt[i][j])
            if i-1>=0 and matrix[i][j] < matrix[i-1][j]:
                inx[i-1][j] -= 1
                if inx[i-1][j] == 0:
                    q1.append([i-1, j])
                cnt[i-1][j] = max(cnt[i-1][j], cnt[i][j]+1)
            if i+1<m and matrix[i][j] < matrix[i+1][j]:
                inx[i+1][j] -= 1
                if inx[i+1][j]==0:
                    q1.append([i+1, j])
                cnt[i+1][j] = max(cnt[i+1][j], cnt[i][j]+1)
            if j-1>=0 and matrix[i][j] < matrix[i][j-1]:
                inx[i][j-1] -= 1
                if inx[i][j-1]==0:
                    q1.append([i, j-1])
                cnt[i][j-1] = max(cnt[i][j-1], cnt[i][j]+1)
            if j+1<n and matrix[i][j] < matrix[i][j+1]:
                inx[i][j+1] -= 1
                if inx[i][j+1]==0:
                    q1.append([i,j+1])
                cnt[i][j+1] = max(cnt[i][j+1], cnt[i][j]+1)
        
        return ret
        
        
                    
        
        

File: 329. Longest Increasing Path in a Matrix.go
func longestIncreasingPath(matrix [][]int) int {
	ret := 0
	m, n := len(matrix), len(matrix[0])
	dp := make([][]int, m)
	for i := range dp {
		dp[i] = make([]int, n)
	}

	max := func(a, b int) int {
		if a < b {
			return b
		}
		return a
	}
	var dfs func(i, j, k, num int)
	dfs = func(i, j, k, num int) {
		if i < 0 || i >= m || j < 0 || j >= n || matrix[i][j] <= num || k <= dp[i][j] {
			return
		}
		ret = max(ret, k)
		dp[i][j] = k

		dfs(i+1, j, k+1, matrix[i][j])
		dfs(i-1, j, k+1, matrix[i][j])
		dfs(i, j+1, k+1, matrix[i][j])
		dfs(i, j-1, k+1, matrix[i][j])
	}
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			dfs(i, j, 1, -1)
		}
	}
	return ret
}